dipy/denoise/localpca.py

```diff
@@ -0,0 +1,111 @@
+import numpy as np
+import scipy as sp
+import scipy.linalg as sla
+from warnings import warn
+
+# Try to get the SVD through direct API to lapack:
+try:
+    from scipy.linalg.lapack import sgesvd as svd
+    svd_args = [1, 0]
+# If you have an older version of scipy, we fall back
+# on the standard scipy SVD API:
+except ImportError:
+    from scipy.linalg import svd
+    svd_args = [False]
+
+
+def localpca(arr, sigma, patch_radius=2):
+    r"""Local PCA-based denoising of diffusion datasets.
+
+    Parameters
+    ----------
+    arr : 4D array
+        Array of data to be denoised. The dimensions are (X, Y, Z, N), where N
+        are the diffusion gradient directions.
+    sigma : float or 3D array
+        Standard deviation of the noise estimated from the data.
+    patch_radius : int
+        The radius of the local patch to be taken around each voxel (in
+        voxels). Default: 2 (denoise in blocks of 5x5x5 voxels).
+
+    Returns
+    -------
+    denoised_arr : 4D array
+        This is the denoised array of the same size as that of the input data.
+
+    References
+    ----------
+    .. [Manjon13] Manjon JV, Coupe P, Concha L, Buades A, Collins DL (2013)
+                  Diffusion Weighted Image Denoising Using Overcomplete Local
+                  PCA. PLoS ONE 8(9): e73021.
+                  https://doi.org/10.1371/journal.pone.0073021
+    """
+    if not arr.ndim == 4:
+        raise ValueError("PCA denoising can only be performed on 4D arrays.",
+                         arr.shape)
+
+    patch_size = 2 * patch_radius + 1
+
+    if patch_size ** 3 < arr.shape[-1]:
+        e_s = "You asked for PCA denoising with a patch_radius of {0} "
+        e_s += "for data with {1} directions. This would result in an "
+        e_s += "ill-conditioned PCA matrix. Please increase the patch_radius."
+        e_s.format(patch_radius, arr.shape[-1])
+        raise ValueError(e_s)
+
+    if isinstance(sigma, np.ndarray):
+        if not sigma.shape == arr.shape[:-1]:
+            e_s = "You provided a sigma array with a shape {0} for data with "
+            e_s = "shape {1}. Please provide a sigma array that matches the "
+            e_s = "spatial dimensions of the data."
+            e_s.format(sigma.shape, arr.shape)
+            raise ValueError(e_s)
+
+    tau = np.ones(arr.shape) * 2.3 * 2.3 * sigma * sigma
```

Comment:
This line give me a crash. For starter,s the error message is not appended, but replaced at each line. The result of using the example as-is is to give you a 3D standard deviation array, but the function accepts a 4D one. Therefore, it is missing a broadcasting (and noise correction) along the last axis.

~~~python

In [8]: 
   ...: t = time()
   ...: sigma = pca_noise_estimate(data, gtab, correct_bias=True, smooth=3)
   ...: print("Sigma estimation time", time() - t)
   ...: 

Sigma estimation time 34.688045024871826

In [9]: 

In [9]: 
   ...: 
   ...: t = time()
   ...: 
   ...: denoised_arr = localpca(data, sigma=sigma, patch_radius=2)
   ...: 
   ...: print("Time taken for local PCA (slow)", -t + time())
   ...: 

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-9945762433a7> in <module>()
      3 t = time()
      4 
----> 5 denoised_arr = localpca(data, sigma=sigma, patch_radius=2)
      6 
      7 print("Time taken for local PCA (slow)", -t + time())

/home/samuel/Bureau/dipy/dipy/denoise/localpca.py in localpca(arr, sigma, patch_radius)
     62             raise ValueError(e_s)
     63 
---> 64     tau = np.ones(arr.shape) * 2.3 * 2.3 * sigma * sigma
     65 
     66     # declare arrays for theta and thetax

ValueError: operands could not be broadcast together with shapes (192,192,96,41) (192,192,96) 
~~~

I am currently making sure it still gives erroneous results with the latest changes.